### PR TITLE
Created converteIfcPointOnCurve

### LINKED ISSUE #359: [TODO] IfcPointOnCurve
**Location in code**
PlacementConverter - around line 100.

**Additional context**
- Docu: https://standards.buildingsmart.org/IFC/DEV/IFC4_3/RC2/HTML/link/ifcpointoncurve.htm
- after the introduction of `GetPointOnCurve` this shouldn't be hard.

### PR DESCRIPTION
**Pull request by SamuilsRulovs**
Fixes #359

### CODE REVIEW

File: Core/src/IfcGeometryConverter/CurveConverter.h [hunk 1]
```diff
@@ -2114,13 +2114,29 @@ namespace OpenInfraPlatform {
 					}
 				}
 
+				/**********************************************************************************************/
+				/*! \brief Calculates a trimming point on the curve using \c IfcCartesianPoint.
+				* \tparam TCurve					A type of the curve.
+				* \param[in] curve					A pointer to data from the curve.
+				* \param[in] cartesianPoint			A pointer to data from \c IfcCartesianPoint.
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/CurveConverter.h]: this doesn't fit the actual function

File: Core/src/IfcGeometryConverter/PlacementConverter.h [hunk 2]
```diff
@@ -153,6 +157,22 @@ namespace OpenInfraPlatform {
...
 
+				/*! \brief Converts \c IfcPointOnCurve to a 3D vector.
+				 *
+				 * \param[in]	pointOnCurve			\c IfcPointOnCurve entity to be interpreted.
+				 *
+				 * \return								Calculated 3D vector.
+				 */
+				carve::geom::vector<3> convertIfcPointOnCurve(
+					const EXPRESSReference<typename IfcEntityTypesT::IfcPointOnCurve>& pointOnCurve
+				) const throw(...) {
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/PlacementConverter.h]:
```suggestion
				) const noexpect(false) {
```

We're no longer using `throw(...)` - use this with all *new*  functions you're writting.

File: Core/src/IfcGeometryConverter/PlacementConverter.h [hunk 3]
```diff
@@ -153,6 +157,22 @@ namespace OpenInfraPlatform {
...
+					std::shared_ptr<PlacementConverterT<IfcEntityTypesT>> placementConverter
+						= std::make_shared<PlacementConverterT<IfcEntityTypesT>>(this->GeomSettings(), this->UnitConvert());
+
+					CurveConverterT<IfcEntityTypesT> gridConv(this->GeomSettings(), this->UnitConvert(), placementConverter);
+					return gridConv.getPointOnCurve<typename IfcEntityTypesT::IfcCurve>(pointOnCurve->BasisCurve, pointOnCurve->PointParameter);
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/PlacementConverter.h]: how come this is called `gridConv`?

File: Core/src/IfcGeometryConverter/PlacementConverter.h [hunk 4]
```diff
@@ -100,7 +101,10 @@ namespace OpenInfraPlatform {
 					//TODO this is really, really bad programming - this needs to be done more nicely!!
 #endif
 
-					// IfcPointOnCurve & IfcPointOnSurface are not supported
+					// IfcPointOnCurve 
+					if (point.isOfType<typename IfcEntityTypesT::IfcPointOnCurve>())
+						return convertIfcPointOnCurve(point.as<typename IfcEntityTypesT::IfcPointOnCurve>());
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/PlacementConverter.h]: templated function should be called with the `template` keyword, e.g. `.template as<...`

File: Core/src/IfcGeometryConverter/PlacementConverter.h [hunk 5]
```diff
@@ -28,6 +28,7 @@
 
 #include "GeomUtils.h"
 #include "ConverterBase.h"
+#include "CurveConverter.h"
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/PlacementConverter.h]: how come it worked before for the grid placement conversion without the include of the curve converter header?

**Review by pjanck — COMMENTED** [review]
I'd assume that some calls to the `getPointOnCurve` function need revision as well to account for the introduced template parameter.

File: Core/src/IfcGeometryConverter/PlacementConverter.h (hunk 5 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/PlacementConverter.h]: Yes, it is redundant, I have added it while trying no include getPointOnCurve function. Resolved: 
795b117

File: Core/src/IfcGeometryConverter/PlacementConverter.h (hunk 4 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/PlacementConverter.h]: Resolved: 795b117

File: Core/src/IfcGeometryConverter/PlacementConverter.h (hunk 2 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/PlacementConverter.h]: Resolved: 795b117

File: Core/src/IfcGeometryConverter/CurveConverter.h (hunk 1 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/CurveConverter.h]: Resolved: 795b117

File: Core/src/IfcGeometryConverter/PlacementConverter.h (hunk 3 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/PlacementConverter.h]: Resolved: 37bb3af

**Comment by SamuilsRulovs** [comment]
> I'd assume that some calls to the `getPointOnCurve` function need revision as well to account for the introduced template parameter.

I guess I found all places, which should be revised. These changes are in commit 795b117

File: Core/src/IfcGeometryConverter/PlacementConverter.h [hunk 6]
```diff
@@ -103,7 +102,7 @@ namespace OpenInfraPlatform {
 
 					// IfcPointOnCurve 
 					if (point.isOfType<typename IfcEntityTypesT::IfcPointOnCurve>())
-						return convertIfcPointOnCurve(point.as<typename IfcEntityTypesT::IfcPointOnCurve>());
+						return convertIfcPointOnCurve(point.template as<typename IfcEntityTypesT::IfcPointOnCurve>());
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/PlacementConverter.h]: `isOfType` is also a templated function :)

File: Core/src/IfcGeometryConverter/PlacementConverter.h [hunk 7]
```diff
@@ -165,12 +164,12 @@ namespace OpenInfraPlatform {
...
 					const EXPRESSReference<typename IfcEntityTypesT::IfcPointOnCurve>& pointOnCurve
-				) const throw(...) {
+				) const noexcept(false) {
 					std::shared_ptr<PlacementConverterT<IfcEntityTypesT>> placementConverter
 						= std::make_shared<PlacementConverterT<IfcEntityTypesT>>(this->GeomSettings(), this->UnitConvert());
 
-					CurveConverterT<IfcEntityTypesT> gridConv(this->GeomSettings(), this->UnitConvert(), placementConverter);
-					return gridConv.getPointOnCurve<typename IfcEntityTypesT::IfcCurve>(pointOnCurve->BasisCurve, pointOnCurve->PointParameter);
+					CurveConverterT<IfcEntityTypesT> getPoint(this->GeomSettings(), this->UnitConvert(), placementConverter);
+					return getPoint.getPointOnCurve<typename IfcEntityTypesT::IfcCurve>(pointOnCurve->BasisCurve, pointOnCurve->PointParameter);
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/PlacementConverter.h]: Not really a better choice. I'd propose `curveConv` since it is a curve converter :)

**Review by pjanck — CHANGES_REQUESTED** [review]

File: Core/src/IfcGeometryConverter/PlacementConverter.h (hunk 7 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/PlacementConverter.h]: Resolved: 0a192a9

File: Core/src/IfcGeometryConverter/PlacementConverter.h (hunk 6 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/PlacementConverter.h]: Resolved: 0a192a9

**Review by pjanck — APPROVED** [review]

**Pull request merged**